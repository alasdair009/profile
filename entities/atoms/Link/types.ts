import { HTMLAttributes, ReactNode } from "react";

export type LinkVariant = "large" | "regular";

export type LinkProps = {
  /**
   * Visual variant of the link.
   */
  variant?: LinkVariant;
  /**
   * Content of the link.
   */
  children?: ReactNode;
  /**
   * Url or path to link to.
   */
  href: string;
  /**
   * Display an iframe preview of the href
   */
  hoverFrame?: boolean;
  /**
   * Number of lines the text is restricted to.
   */
  lines?: number;
  /**
   * How the anchor impacts the view.
   */
  target?: "_blank" | "_self" | "_parent";
} & HTMLAttributes<HTMLAnchorElement>;
