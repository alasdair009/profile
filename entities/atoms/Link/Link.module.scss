@use "../../../app/styles/colors";
@use "../../../app/styles/animation";
@use "../../../app/styles/dimensions";
@use "../../../app/styles/typography";
@use "../../../app/styles/lineClamp";
@use "../../../app/styles/helpers/rem";

@function generatePolygonLine($yOffset: 0) {
  @return polygon(
    0 0,
    30% #{$yOffset * 0.9%},
    50% #{$yOffset * 1%},
    70% #{$yOffset * 0.9%},
    100% 0,
    100% 25%,
    70% #{25 + $yOffset * 0.9%},
    50% #{25 + $yOffset * 1%},
    30% #{25 + $yOffset * 0.9%},
    0 25%
  );
}

@keyframes pressDown {
  from {
    clip-path: generatePolygonLine(0);
  }
  to {
    clip-path: generatePolygonLine(50);
  }
}

@keyframes release {
  0% {
    clip-path: generatePolygonLine(50);
  }
  60% {
    clip-path: generatePolygonLine(-30);
  }
  80% {
    clip-path: generatePolygonLine(20);
  }
  100% {
    clip-path: generatePolygonLine(0);
  }
}

.root {
  align-items: center;
  color: colors.$colorWhiteGhost;
  display: inline-flex;
  font-size: inherit;
  font-weight: map-get(typography.$fontWeights, bold);
  max-width: rem.toRem(map-get(dimensions.$sizes, s512));
  position: relative;
  text-decoration: none;
  word-break: break-word;

  &:hover {
    filter: brightness(1.3);

    &::after {
      animation: pressDown animation.animationDuration(0.1);
      animation-fill-mode: forwards;
    }

    .hoverIFrameWrapper {
      display: block;
    }
  }

  &:visited,
  &:active {
    color: colors.$colorWhiteGhost;
  }

  &::after {
    animation: release animation.animationDuration(0.2);
    animation-fill-mode: forwards;
    background: colors.$colorWhiteGhost;
    bottom: calc(#{rem.toRem(7)} * -1);
    clip-path: generatePolygonLine();
    content: "";
    height: rem.toRem(map-get(dimensions.$sizes, s8));
    left: 0;
    position: absolute;
    width: 100%;
  }

  &Clamp {
    @include lineClamp.lineClamp;
  }

  &.Large {
    color: colors.$colorGreenGrass;
    font-size: rem.toRem(map-get(typography.$fontSizes, large));
    padding-bottom: calc(rem.toRem(map-get(dimensions.$sizes, s16)) + 2px);
    text-transform: uppercase;

    &:visited,
    &:active {
      color: colors.$colorGreenGrass;
    }

    &::after {
      bottom: 0;
      height: rem.toRem(map-get(dimensions.$sizes, s16));
    }
  }
}

.newTabIcon {
  display: inline-block;
  height: rem.toRem(map-get(dimensions.$sizes, s16));
  margin-left: rem.toRem(map-get(dimensions.$sizes, s4));
  width: rem.toRem(map-get(dimensions.$sizes, s16));
}

.hoverIFrameWrapper {
  aspect-ratio: 16 / 9;
  background: colors.$colorGreyDark;
  border: rem.toRem(map-get(dimensions.$sizes, s8)) solid colors.$colorGreyDark;
  box-shadow: 0 0 rem.toRem(2) rem.toRem(2) colors.$colorblackEvil;
  display: none;
  height: calc(
    rem.toRem(map-get(dimensions.$sizes, s128)) +
      rem.toRem(map-get(dimensions.$sizes, s16))
  );
  overflow: hidden;
  position: absolute;
  left: 50%;
  top: 100%;
  transform: translate(-50%, 0);
  transition: all animation.animationDuration(0.2);
  user-select: none;
}

.hoverIFrame {
  aspect-ratio: unset;
  height: calc(1000% - rem.toRem(map-get(dimensions.$sizes, s16)));
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%) scale(0.5);
  width: calc(1000% - rem.toRem(map-get(dimensions.$sizes, s16)));
}
