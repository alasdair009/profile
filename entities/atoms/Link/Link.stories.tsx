import { Link } from "./Link";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof Link> = {
  component: Link,
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
  },
  args: {
    href: "#",
    children: "Go to page",
  },
};
export default meta;

export const Default: StoryObj<typeof Link> = {};
