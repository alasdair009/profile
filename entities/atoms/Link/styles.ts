"use client";
import Link from "next/link";
import styled, { keyframes } from "styled-components";
import { colors, fontSizes, fontWeights, sizes } from "@/entities";
import { LinkVariant } from "./types";
import Image from "next/image";

const generatePolygonLine = (yOffset = 0) => {
  return `polygon(0 0, 30% ${yOffset * 0.9}%, 50% ${yOffset}%, 70% ${
    yOffset * 0.9
  }%, 100% 0, 100% 25%, 70% ${25 + yOffset * 0.9}%, 50% ${25 + yOffset}%, 30% ${
    25 + yOffset * 0.9
  }%, 0 25%)`;
};

const pressDown = keyframes`
  from {
    clip-path: ${generatePolygonLine(0)};
  }
  to {
    clip-path: ${generatePolygonLine(50)};
  }
`;

const release = keyframes`
  0% {
    clip-path: ${generatePolygonLine(50)};
  }
  60% {
    clip-path: ${generatePolygonLine(-30)};
  }
  80% {
    clip-path: ${generatePolygonLine(20)};
  }
  100% {
    clip-path: ${generatePolygonLine(0)};
  }
`;

export const Root = styled(Link)<{ $variant: LinkVariant }>`
  color: ${({ $variant }) =>
    $variant === "large" ? colors.greenGrass : colors.whiteGhost};
  font-size: ${({ $variant }) =>
    $variant === "large" ? fontSizes.large.rem : "inherit"};
  font-weight: ${fontWeights.bold};
  overflow: hidden;
  padding-bottom: ${({ $variant }) =>
    $variant === "large"
      ? `calc(${sizes.s16.rem} + 2px)`
      : `calc(${sizes.s8.rem} + 2px)`};
  position: relative;
  text-decoration: none;
  text-transform: ${({ $variant }) =>
    $variant === "large" ? "uppercase" : null};
  word-break: break-word;

  &:hover {
    filter: brightness(1.3);

    &::after {
      animation: ${pressDown} 0.1s;
      animation-fill-mode: forwards;
    }
  }

  &:visited,
  &:active {
    color: ${({ $variant }) =>
      $variant === "large" ? colors.greenGrass : colors.whiteGhost};
  }

  &::after {
    animation: ${release} 0.2s;
    animation-fill-mode: forwards;
    background: ${colors.whiteGhost};
    bottom: 0;
    clip-path: ${() => generatePolygonLine()};
    content: "";
    height: ${({ $variant }) =>
      $variant === "large" ? sizes.s16.rem : sizes.s8.rem};
    left: 0;
    position: absolute;
    width: 100%;
  }
`;

export const NewTabIcon = styled(Image)`
  display: inline-block;
  height: ${sizes.s16.rem};
  margin-left: ${sizes.s4.rem};
  width: ${sizes.s16.rem};
`;
