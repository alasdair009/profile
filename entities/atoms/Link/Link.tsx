import { NewTabIcon, StaticLink, TransitionLink } from "./styles";
import type { LinkProps } from "./types";
import { isExternalDomain } from "@/lib/domains";
import newTabIcon from "../../assets/new-tab-icon.svg";

/**
 * Link to a new path or url.
 */
export function Link({
  variant = "regular",
  children,
  href,
  ...rest
}: LinkProps) {
  const isExternalURL = isExternalDomain(href);

  if (process.env.ALLOW_TRANSITIONS) {
    return (
      <TransitionLink
        $variant={variant}
        href={href}
        target={isExternalURL ? "_blank" : "_self"}
        rel={isExternalURL ? "noopener" : undefined}
        data-testid={Link.name}
        {...rest}
      >
        {children ? children : href}
        {isExternalURL && (
          <NewTabIcon
            src={newTabIcon}
            title="Third party site, opens in a new tab."
            alt="Icon to represent this opens in a new tab"
          />
        )}
      </TransitionLink>
    );
  }

  return (
    <StaticLink
      $variant={variant}
      href={href}
      target={isExternalURL ? "_blank" : "_self"}
      rel={isExternalURL ? "noopener" : undefined}
      data-testid={Link.name}
      {...rest}
    >
      {children ? children : href}
      {isExternalURL && (
        <NewTabIcon
          src={newTabIcon}
          title="Third party site, opens in a new tab."
          alt="Icon to represent this opens in a new tab"
        />
      )}
    </StaticLink>
  );
}
