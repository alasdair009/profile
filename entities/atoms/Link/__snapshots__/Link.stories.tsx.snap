// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Link Default play-test 1`] = `
<a target="_self"
   data-testid="Link"
   class="sc-oeqTF cFwkOt"
   href="#"
>
  Go to page
</a>
`;

exports[`atoms/Link HoverFrame play-test 1`] = `
<a target="_self"
   data-testid="Link"
   class="sc-oeqTF cFwkOt"
   href="https://www.alasdairmacrae.co.uk"
>
  Go to page
  <div class="sc-drnuxz gQjKNO">
    <iframe src="https://www.alasdairmacrae.co.uk"
            class="sc-dENhDJ hcyAeG sc-hDcvty jDAvkM"
            allowfullscreen
            data-testid="IFrame"
    >
    </iframe>
  </div>
</a>
`;
